Test framework migration

niveditn opened the pull request:
**Details:**
- Migrates storybook based test framwork to jest+enzyme, with refactors
- Lifecycle and event based tests added
- Coverage increased from ~50% to 84%
- For #20, added codecov integration
- Bumps `imagekit-javascript` to `^1.3.6`

niveditn committed:
remove storybook and old test files

niveditn committed:
add new test files

niveditn committed:
add coverage files

niveditn committed:
migrate IKContext, IKUpload and transform snapshots to jest

niveditn committed:
refactor all tests and add new ones

niveditn committed:
update coverage

niveditn committed:
commit IKUpload event tests with stubs

niveditn committed:
update coverage

niveditn committed:
IKImage: add new cases to increase coverage

niveditn committed:
ignore coverage files

imagekitio commented on package.json:
```diff
@@ -36,24 +42,23 @@
     "imagekit-javascript": "^1.3.5",
-    "prop-types": "^15.7.2"
+    "prop-types": "^15.7.2",
+    "react": "^16.13.1",
```
`react` and `react-dom` should be in peer dependency

niveditn replied:
Done. They were also needed in `devDependencies` for test cases to run successfully, so added there as well.

imagekitio commented on src/test/IKImage.test.js:
```diff
@@ -0,0 +1,191 @@
+      const ikImage = shallow(<IKImage urlEndpoint={urlEndpoint} path={relativePath} alt={'some text here'} />);
+      // <img alt="" src="http://ik.imagekit.io/test_imagekit_id/default-image.jpg?ik-sdk-version=react-1.0.7"/>
+
+      expect(ikImage.find('img').prop('src')).toEqual(`${absolutePath}?${global.SDK_VERSION}`);
```
For better code readability, instead of using the  stored `absolutePath`, here we should derive `absolutePath` 
 by replacing `absolutePath` with `${urlEndpoint}/${relativePath}`. I think we can make this change throughout all places where we need to clarify what the final URL is made of.

For example, below, it makes sense to use `absolutePath` so no need to replace.

```
const ikImage = shallow(<IKImage urlEndpoint={urlEndpoint} src={absolutePath} alt={'some text here'} />);
// <img alt="abc" src="http://ik.imagekit.io/test_imagekit_id/default-image.jpg
// ?ik-sdk-version=react-1.0.7"/>

expect(ikImage.find('img').prop('src')).toEqual(`${absolutePath}?${global.SDK_VERSION}`);
expect(ikImage.find('img').prop('alt')).toEqual('some text here');
```

niveditn replied:
Done throughout the test files.

imagekitio commented on src/test/IKImageTransformation.test.js:
```diff
@@ -0,0 +1,135 @@
+import React from 'react';
```
if the test filename and component name matches then `IKImageTransformation` makes no sense. We can either use filenames to show what test cases are about or stick with component name convention and club test cases together.

niveditn replied:
These tests have been moved into `IKImage.test.js`

imagekitio commented on src/test/IKImageTransformation.test.js:
```diff
@@ -0,0 +1,135 @@
+  });
+});
+
+describe('IKImage transformations', () => {
```
The group name `IKImage transformations` is not clarifying what it consists of. It should be clear what each test case/group is doing just by looking at the output.

```
PASS src/test/IKUpload.test.js
  IKUpload
    Success case
      ✓ should successfully upload file (54 ms)
    Failure cases
      ✓ should call onError for missing publicKey (4 ms)
      ✓ should call onError for missing urlEndpoint (2 ms)
      ✓ should call onError for missing authenticationEndpoint (3 ms)
      ✓ should call onError in case server returns error response (4 ms)
    General cases
      ✓ should call onChange when file input changes (3 ms)
  IKUpload snapshots
    ✓ with authentication (2 ms)
    ✓ with all the props (2 ms)
    ✓ without fileName prop (1 ms)
    ✓ without authentication (2 ms)

PASS src/test/IKImage.test.js
  IKImage snapshots
    ✓ path with transformations and lqip (2 ms)
    ✓ nested path with lqip (3 ms)
    absolute image path
      ✓ src with alt attribute (7 ms)
      ✓ src with query parameters (2 ms)
      ✓ src with transformation (1 ms)
      ✓ src with lqip (3 ms)
      ✓ src with transformation and lqip (2 ms)
    relative image path
      ✓ path with alt attribute (3 ms)
      ✓ path with query parameters (1 ms)
      ✓ path having leading slashes (2 ms)
      ✓ path with url endpoint having trailing slashes (2 ms)
      ✓ path with lqip (3 ms)
      ✓ path with transformations (2 ms)

PASS src/test/IKImageTransformation.test.js
  IKImage transformation snapshots
    ✓ single transformation (3 ms)
    ✓ transformation position as query (2 ms)
    ✓ transformation position as path while using relative image path (2 ms)
    ✓ transformation position as path while using absolute image path (1 ms)
    ✓ chained transformations (2 ms)
    ✓ non-existent transformation (2 ms)
    ✓ non-existent transformation with existing transformation (1 ms)
  IKImage transformations
    ✓ observer disconnects when component unmounts (3 ms)

PASS src/test/IKContext.test.js
  IKContext snapshots
    ✓ with only urlEndpoint (19 ms)
    ✓ with publicKey and urlEndpoint (3 ms)
    ✓ with publicKey, urlEndpoint and authenticationEndpoint (4 ms)
    ✓ overriding urlEndpoint (4 ms)

PASS src/test/ErrorBoundary.test.js
  ErrorBoundary
    ✓ handles error thrown by child component (21 ms)
  ErrorBoundary snapshots
    IKImage
      ✓ missingUrlEndpointFail (70 ms)
      ✓ imageUrlEndpointFail (4 ms)
```

niveditn replied:
Grouping and test case labeling improved for all components.

imagekitio commented on src/test/IKImageTransformation.test.js:
```diff
@@ -0,0 +1,135 @@
+        loading="lazy"
+      />
+    );
+    // spies
```
Test cases for lazyloading are missing.

niveditn replied:
A basic test cased added for lazy load initial stage. Another issue #24 has been created to address event based tests for Lazy Load and LQIP.

imagekitio reviewed (changes requested):
Please add coverage label in README.md like iOS and Android

niveditn committed:
Addressed changes:
- react and react-dom moved to peerDeps and devDeps (needed for tests)
- defined transformURL construction clearly thoughout, replacing absolutePath where needed
- improved test grouping and labels for all components
- removed IKImageTransformations test file
- added basic initial test for lazy loading

niveditn committed:
bump version: imagekit-javascript

niveditn committed:
codecov added

niveditn committed:
codecov added to devDeps

niveditn committed:
add codecov badge

niveditn commented:
@imagekitio Codecov setup done. After this PR is merged, coverage diff will be displayed properly with respect to `master` in subsequent PRs.

niveditn commented:
@imagekitio `imagekit-javascript` version also bumped to `^1.3.6`

niveditn committed:
add test for ImageKitComponent

manu4543 committed:
test cases readability

manu4543 committed:
test case slash fix
